tests/basecli_test.go

```diff
@@ -0,0 +1,222 @@
+	igor     = ACCOUNTS[3]
+)
+
+func gopath() string {
+	return filepath.Join(os.Getenv("GOPATH"), "src", "github.com", "cosmos", "cosmos-sdk")
+}
+
+func whereIsBasecoind() string {
+	return filepath.Join(gopath(), basecoind)
+}
+
+func whereIsBasecli() string {
+	return filepath.Join(gopath(), basecli)
+}
+
+func TestInitBaseCoin(t *testing.T) {
+	clean()
+
+	var err error
+
+	password := "some-random-password"
+	usePassword := exec.Command("echo", password)
+
+	initBasecoind := exec.Command(whereIsBasecoind(), "init", "--home", basecoindDir)
+
+	initBasecoind.Stdin, err = usePassword.StdoutPipe()
+	if err != nil {
+		t.Error(err)
+	}
+
+	initBasecoind.Stdout = os.Stdout
+
+	if err := initBasecoind.Start(); err != nil {
+		t.Error(err)
+	}
+	if err := usePassword.Run(); err != nil {
+		t.Error(err)
+	}
+	if err := initBasecoind.Wait(); err != nil {
+		t.Error(err)
+	}
+
+	if err := makeKeys(); err != nil {
+		t.Error(err)
+	}
+}
+
+func makeKeys() error {
+	var err error
+	for _, acc := range ACCOUNTS {
+		pass := exec.Command("echo", "1234567890")
+		makeKeys := exec.Command(whereIsBasecli(), "keys", "add", acc, "--home", basecliDir)
+
+		makeKeys.Stdin, err = pass.StdoutPipe()
+		if err != nil {
+			return err
+		}
+
+		makeKeys.Stdout = os.Stdout
+		if err := makeKeys.Start(); err != nil {
+			return err
+		}
+		if err := pass.Run(); err != nil {
+			return err
+		}
+
+		if err := makeKeys.Wait(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// these are in the original bash tests
+func TestBaseCliRecover(t *testing.T) {}
+func TestBaseCliShow(t *testing.T)    {}
+
+func _TestSendCoins(t *testing.T) {
+	if err := startServer(); err != nil {
+		t.Error(err)
+	}
+
+	// send some coins
+	// [zr] where dafuq do I get a FROM (oh, use --name)
+
+	sendTo := fmt.Sprintf("--to=%s", bob)
+	sendFrom := fmt.Sprintf("--from=%s", alice)
+
+	cmdOut, err := exec.Command(whereIsBasecli(), "send", sendTo, "--amount=1000mycoin", sendFrom, "--seq=0").Output()
+	if err != nil {
+		t.Error(err)
+	}
+
+	fmt.Printf("sent: %s", string(cmdOut))
+
+}
+
+// expects TestInitBaseCoin to have been run
+func startServer() error {
+	// straight outta https://nathanleclaire.com/blog/2014/12/29/shelled-out-commands-in-golang/
+	cmdName := whereIsBasecoind()
+	cmdArgs := []string{"start", "--home", basecoindDir}
+
+	cmd := exec.Command(cmdName, cmdArgs...)
+	cmdReader, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(cmdReader)
+	go func() {
+		for scanner.Scan() {
+			fmt.Printf("running [basecoind start] %s\n", scanner.Text())
+		}
+	}()
+
+	err = cmd.Start()
+	if err != nil {
+		return err
+	}
+
+	err = cmd.Wait()
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(5 * time.Second)
+
+	return nil
+
+	// TODO return cmd.Process so that we can later do something like:
+	// cmd.Process.Kill()
+	// see: https://stackoverflow.com/questions/11886531/terminating-a-process-started-with-os-exec-in-golang
+}
+
+func clean() {
```

Comment:
I wonder if there is any place here to have a `TestMain` function  (https://golang.org/pkg/testing/#hdr-Main ) which calls `clean()` after running all of the tests? Just a thought

Reply:
++